runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/windmill/WindmillStreamPool.java

```diff
@@ -0,0 +1,178 @@
+   * capacity since the random used to grab a random stream is bound by the size of the list.
+   */
+  @SuppressWarnings("nullness")
+  private static <StreamT extends WindmillServerStub.WindmillStream>
+      List<WindmillStreamPool.StreamData<StreamT>> emptyStreamList(int capacity) {
+    List<WindmillStreamPool.StreamData<StreamT>> streams = new ArrayList<>(capacity);
+    for (int i = 0; i < capacity; i++) {
+      streams.add(null);
+    }
+
+    return streams;
+  }
+
+  @VisibleForTesting
+  static <StreamT extends WindmillServerStub.WindmillStream> WindmillStreamPool<StreamT> forTesting(
+      Duration streamTimeout,
+      Supplier<StreamT> streamSupplier,
+      List<WindmillStreamPool.StreamData<StreamT>> streams,
+      ConcurrentHashMap<StreamT, StreamData<StreamT>> holds) {
+    return new WindmillStreamPool<>(streamTimeout, streamSupplier, streams, holds);
+  }
+
+  /**
+   * Returns a stream for use that may be cached from a previous call. Each call of getStream must
+   * be matched with a call of releaseStream. If the stream has been cached but has timed out and no
+   * longer has any holds, the stream will be closed.
+   *
+   * @implNote Streams are stored in an indexed list where the index is the key to fetch the stream.
+   *     This is so that we can "randomly" fetch a stream for better load balancing.
+   */
+  public StreamT getStream() {
+    int index = ThreadLocalRandom.current().nextInt(streams.size());
+    StreamT validStream;
+    StreamT closeThisStream = null;
+    synchronized (this) {
+      WindmillStreamPool.StreamData<StreamT> existingStreamData = streams.get(index);
+      // There are 3 possible states that can result from fetching the stream from the cache.
+      // 1. Stream doesn't exist create and cache a new one.
+      if (existingStreamData == null) {
+        validStream = createAndCacheStream(index).stream;
+      }
+      // 2. The stream exists, but has timed out and can be marked as closed. The timed out stream
+      // is not returned (a new one is created and returned here) and evicted from the cache. Every
+      // call to getStream(), is matched with a call to releaseStream(), so the stream will
+      // eventually be closed there if not here.
+      else if (existingStreamData.hasTimedOut(streamTimeout) && --existingStreamData.holds == 0) {
+        holds.remove(existingStreamData.stream);
+        // Mark the stream as to be closed outside the lock.
+        closeThisStream = existingStreamData.stream;
+        // Cached stream is marked as to be closed, create and cache a new stream at the closed
+        // stream's index.
+        validStream = createAndCacheStream(index).stream;
+      } else {
+        // 3. The stream exists and is in a valid state.
+        existingStreamData.holds++;
+        validStream = existingStreamData.stream;
+      }
+    }
+
+    if (closeThisStream != null) {
```

Comment:
perhaps this would be safer in a finally block in case createAndCacheStream throws an exception

Reply:
can we close it within the lock?

Reply:
it would be better to keep the close outside the lock but that seems like it coudl be done with

try {
  synchronized {
    if (...) {
      closeThisStream = ...
    }
  }
} finally {
  if (closeThisStream) closeThisStream.close();
}

Reply:
done.